Mixpanel Increment Property Action

davidbielik opened the pull request:
Followup to https://github.com/segmentio/action-destinations/pull/1403

Adds Mixpanel support to [increment profile properties](https://developer.mixpanel.com/reference/profile-numerical-add) via the $add API field.


## Testing
- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

davidbielik commented:
I can't get the `/dev-center/action-tester` to work properly. The fields show up as expected in the payload, but I think the UI is showing an array of objects instead of just a simple object?

<img width="515" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/2aef4cda-a35d-482a-8683-65eefab79487">

<img width="1025" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/4dcc35ba-921e-4c1e-abfd-2e5c369dc52f">

<img width="620" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/12b7fa22-04bc-4bcb-aeba-72b54a887da9">

joe-ayoub-segment commented:
Thanks for raising this PR. I'll schedule it for review.

joe-ayoub-segment reviewed (changes requested):
Hi David, 
I've got some questions about the design of this change. Would it be possible to meet and discuss please? Please see the other comment for meeting invite instructions. 

Kind regards,
Joe

davidbielik commented on packages/browser-destinations/destinations/wisepops/src/generated-types.ts:
```diff
@@ -2,7 +2,7 @@
 export interface Settings {
   /**
-   * The identifier of your Wisepops' website. You can find it in [your setup code on Wisepops](https://app.wisepops.com/f/settings/websites).
+   * The identifier of your Wisepops' website. You can find it in [your setup code on Wisepops](https://id.wisepops.com/r/id/workspaces/_workspaceId_/settings/setup-code).
```
woops this might have been a bad merge. i'll try and fix this

davidbielik replied:
oh - actually this was from running the `./bin/run generate:types` command. It must have not been run when someone else updated those comments

davidbielik commented:
Hi @joe-ayoub-segment I've updated to use the `identifyUser` action, but seeing strange behavior in the action tester.

Even though I have it listed as an object, when I choose the object from `traits` it is converted to a string in the resolved payload:

<img width="1498" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/913c9285-96fd-4b23-9a82-44efa12ad062">

Any ideas what's going wrong here?

joe-ayoub-segment commented:
Hi @davidbielik I tested it out. 

Seemed to work fine when I provided paths for each individual key:value. 
![image](https://github.com/segmentio/action-destinations/assets/45374896/b355a164-a1fa-4a46-a548-751c851f2b1c)

joe-ayoub-segment commented:
It also worked if I changed the default path to remove the dollar symbol. 

![image](https://github.com/segmentio/action-destinations/assets/45374896/cdc851f3-59fa-4a1b-97e0-390668d6ca09)

joe-ayoub-segment commented:
So I recommend changing the default path from this: 

![image](https://github.com/segmentio/action-destinations/assets/45374896/9c1534d0-0e5d-4007-b928-c62325a88c07)


To something like this: 

![image](https://github.com/segmentio/action-destinations/assets/45374896/063adac8-3e7d-41ca-a42a-45765e405bdb)


Take the dollar symbol out, and possibly rename from "add" to something more user friendly?

joe-ayoub-segment commented:
Note: Asked @spjtls9 to approve this PR.

davidbielik commented:
Thanks @joe-ayoub-segment 

After further testing it doesn't seem like Mixpanel can support this as easily as I thought.

You can't combine `$add` and `$set` in the same `/engage` request:
<img width="620" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/b305de6b-6075-4f17-b9d4-91c13f456c96">

And when I split it out into two parallel requests, only one seems to be set on the profile, maybe some race condition for new users.

Is it possible to loop in someone from Mixpanel? I was hoping I could add the `$add`, `$union`, `$append`, and `$remove` operations.

spjtls9 commented:
> can't combine `$add` and `$set` in the same `/engage` request:

Hi @davidbielik I am from Mixpanel 👋 

I have to check with service owners internally to see if this is by design. Will get back to you when I hear back.

joe-ayoub-segment commented:
hi @spjtls9 - just following up on this to see if you've been able to unblock @davidbielik. 
Best regards,
Joe

joe-ayoub-segment commented:
Hi @spjtls9 and @davidbielik - any update on this PR? Shall I close it for now?

davidbielik commented:
> > Hi @spjtls9 and @davidbielik - any update on this PR? Shall I close it for now?
> 
> Hi @davidbielik I got some insights from my engineers.
> 
> > When executing two /engage calls, one with $add and one with $set serially, only one works
> 
> * They should both "work," but we make no guarantees about ordering in the engage API (this is a known gap). This means that if we process the $add first, we'll overwrite the result with the $set. Because of this, we recommend any customer implementing non-idempotent operations (append, union, add)
> * In short, for best practice, we only recommend `set` for `/engage`

So does this mean they should be done in separate calls? i'm seeing that only one of the requests are processed. even if they don't have conflicting/overlapping fields.

joe-ayoub-segment commented:
hi @davidbielik and @spjtls9 - just checking in on this. If you are unsure of a way forward and need to think about it can we convert this to a draft PR please?

spjtls9 commented:
> hi @davidbielik and @spjtls9 - just checking in on this. If you are unsure of a way forward and need to think about it can we convert this to a draft PR please?

Per recommendation from Infra, we prefer that only `set` is implemented.

davidbielik commented:
Hi @spjtls9 how should customers increment properties via API? There's nothing in your public docs that say it shouldn't be used, so unsure what your recommendation is. `$add` is supported in your client-side libraries.

If i separate it out into its own segment Action, would that be ok?

spjtls9 commented:
> Hi @spjtls9 how should customers increment properties via API? There's nothing in your public docs that say it shouldn't be used, so unsure what your recommendation is. `$add` is supported in your client-side libraries.
> 
Ideally, we'd like a `set` with the result of the add, but that wouldn't work without the current value.

> If i separate it out into its own segment Action, would that be ok?

For non-idempotent operations (add, union, append), yes -- this sounds like the best option for now.

davidbielik commented:
Thanks @spjtls9 

@joe-ayoub-segment - should I switch this back to its own action? or can we add a note somewhere to say this field should not be used alongside other updates?

joe-ayoub-segment commented:
Hi @davidbielik and @spjtls9 , 

Following up on this PR - is it something you'd like to progress? I'd be happy to get on a call to clarify anything?
http://calendly.com/joe_ayoub/integration-workshop

Best regards,
Joe

davidbielik commented:
Hi @joe-ayoub-segment sorry for the delay. I'll try to get a PR up today with the new action.

davidbielik commented on packages/destination-actions/src/destinations/mixpanel/identifyUser/index.ts:
```diff
@@ -114,6 +114,7 @@ const action: ActionDefinition<Settings, Payload> = {
       })
       responses.push(engageResponse)
     }
+
```
```suggestion
```

davidbielik commented:
QA Notes

**Request Payload**
<img width="1511" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/5d141c93-6518-4d89-bc73-b330baeba0be">


**Request Response**
<img width="1512" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/f8036e5e-f73f-4d35-86bb-9bc40747e85c">


**Profile Created with attributes**
<img width="454" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/d6e4967c-b1da-4145-a1bc-34e8bd864dd3">

**Profile updated with new values**
<img width="450" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/50d51c4b-fbae-4cfd-95c8-4c807c8bc0cf">

**Tests pass**
<img width="887" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/ce06080a-d5c5-4391-8e44-f5b40c6e3cd7">

joe-ayoub-segment commented:
hi @davidbielik , looks like you've made some progress. Let me know when you want me to take a look.

davidbielik commented:
Hi @joe-ayoub-segment it's ready for review. Thanks!

joe-ayoub-segment commented on packages/destination-actions/src/destinations/mixpanel/incrementProperties/index.ts:
```diff
@@ -0,0 +1,89 @@
+  },
+
+  perform: async (request, { payload, settings }) => {
+    if (!settings.projectToken) {
+      throw new IntegrationError('Missing project token', 'Missing required field', 400)
+    }
```
Not needed as projectToken is a required field. But OK to leave it I guess.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/mixpanel/incrementProperties/index.ts:
```diff
@@ -0,0 +1,89 @@
+      responses.push(response)
+    }
+
+    return Promise.all(responses)
```
I wonder should we put a hard limit on the number of requests which can be made? Say 10 or 15?
If you do decide on a limit, let's throw an error if the customer provides more than number of key:value pairs.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/mixpanel/incrementProperties/index.ts:
```diff
@@ -0,0 +1,89 @@
+      }
+      data.$add = {}
+
+      for (const key of Object.keys(payload.increment)) {
+        const value = payload.increment[key]
+        if (typeof value === 'string' || typeof value === 'number') {
+          if (!isNaN(+value)) {
+            data.$add[key] = +value
+          }
+        }
+      }
```
Can we throw an error if any of the values are not numeric? Best to provide feedback to the customer by doing this.

joe-ayoub-segment commented:
Hi @davidbielik , please see responses / recommendations inline. Do you know if the unit tests pass?

davidbielik commented:
> Hi @davidbielik , please see responses / recommendations inline. Do you know if the unit tests pass?

Hi @joe-ayoub-segment addressed your comments (thanks!) and yes tests passed:

<img width="1195" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/b25e4e63-d23c-47b8-8cbd-3bee7a055017">

davidbielik commented:
> Hi @davidbielik , please see responses / recommendations inline. Do you know if the unit tests pass?

Hi @joe-ayoub-segment I updated the PR, let me know if there's any other feedback - thanks!

joe-ayoub-segment commented:
Follows up again with Mixpanel. 
![image](https://github.com/segmentio/action-destinations/assets/45374896/b102bf59-ceb9-4ace-bc53-0dec64753b91)

spjtls9 reviewed (changes requested):
Thanks for this change -- I am double-checking internally to make sure everything in this PR looks good. Holding off approval until I hear back from them.

spjtls9 reviewed (approved):
I recommend increasing the limit to 100 -- 20 seems a bit too small.

Thank you for the PR!

joe-ayoub-segment commented on packages/destination-actions/src/destinations/mixpanel/incrementProperties/index.ts:
```diff
@@ -0,0 +1,100 @@
+    if (payload.increment && Object.keys(payload.increment).length > 0) {
+      const keys = Object.keys(payload.increment)
+      if (keys.length > 20) {
+        throw new IntegrationError(
```
hi  @davidbielik and @spjtls9 , PayloadValidationError might be more appropriate here.

joe-ayoub-segment commented:
Hi @davidbielik and @spjtls9 , 

If I'm not mistaken, @spjtls9 is good with this PR as long as a couple of comments are taken on board. Is that correct @spjtls9 ?

Best regards,
Joe

davidbielik commented on packages/destination-actions/src/destinations/mixpanel/incrementProperties/index.ts:
```diff
@@ -0,0 +1,100 @@
+
+    if (payload.increment && Object.keys(payload.increment).length > 0) {
+      const keys = Object.keys(payload.increment)
+      if (keys.length > 20) {
```
```suggestion
      if (keys.length > 100) {
```

davidbielik commented:
@joe-ayoub-segment updated to `PayloadValidationError` and increased limit from 20 to 100.

joe-ayoub-segment commented:
Note: I've asked @spjtls9 on Slack if she would be willing to help me test/validate this PR in our internal Staging environment before deploying.  Awaiting a response.

spjtls9 commented on packages/destination-actions/src/destinations/mixpanel/incrementProperties/index.ts:
```diff
@@ -0,0 +1,98 @@
+        }
+      }
+
+      const response = request(`${apiServerUrl}/engage`, {
+        method: 'post',
+        body: new URLSearchParams({ data: JSON.stringify(data) })
+      })
```
URL for `increment` has changed https://developer.mixpanel.com/reference/profile-numerical-add

spjtls9 commented on packages/destination-actions/src/destinations/mixpanel/incrementProperties/index.ts:
```diff
@@ -0,0 +1,98 @@
+      const keys = Object.keys(payload.increment)
+      if (keys.length > 100) {
+        throw new PayloadValidationError(
+          'Exceeded maximum of 20 properties for increment call'
```
Update comment to match the logic

joe-ayoub-segment commented on packages/destination-actions/src/destinations/mixpanel/incrementProperties/index.ts:
```diff
@@ -0,0 +1,98 @@
+
+      for (const key of keys) {
+        const value = payload.increment[key]
+        if (typeof value === 'string' || typeof value === 'number') {
```
Should only allow type === number

joe-ayoub-segment commented:
This PR is too stale so I will copy details over to a new PR - and deploy it.

joe-ayoub-segment commented:
Hi @davidbielik this PR has been deployed (the code from it) via https://github.com/segmentio/action-destinations/pull/1802

Could you confirm that your change looks good to you please?

davidbielik commented:
Hi @joe-ayoub-segment it looks like the prod UI is behaving differently than the local tester i previously used. Or am I not using it correctly? I am trying to hard-code a string as the key but it's showing reversed?

<img width="869" alt="image" src="https://github.com/segmentio/action-destinations/assets/7410770/6f8c4d11-6444-44fe-b7ec-42dc3d8efd9e">

joe-ayoub-segment commented:
hi @davidbielik 
That's OK - the Prod UI is a little different. The position of the fields is reversed for some reason! It should still work as before though, so you should be able to do an end-to-end test. 
Happy to help if needed - please let me know. 
